hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -76,68 +73,75 @@ public MapQueryEngineImpl(MapServiceContext mapServiceContext) {
         this.queryResultSizeLimiter = new QueryResultSizeLimiter(mapServiceContext, logger);
         this.operationService = nodeEngine.getOperationService();
         this.clusterService = nodeEngine.getClusterService();
-        this.queryDispatcher = new MapQueryDispatcher(mapServiceContext);
+        this.queryDispatcher = new QueryDispatcher(mapServiceContext);
+        this.resultProcessorRegistry = mapServiceContext.getResultProcessorRegistry();
     }
 
-    // query thread first, fallback to partition thread
     @Override
-    public QueryResult runQueryOnLocalPartitions(String mapName, Predicate predicate, IterationType iterationType) {
+    public Result execute(Query query, Target target) {
+        IterationType retrievalIterationType = getRetrievalIterationType(query.getPredicate(), query.getIterationType());
+        Query adjustedQuery = Query.of(query).iterationType(retrievalIterationType).build();
+        if (target.isTargetAllNodes()) {
+            return runQueryOnAllPartitions(adjustedQuery);
+        } else if (target.isTargetLocalNode()) {
+            return runQueryOnLocalPartitions(adjustedQuery);
+        } else if (target.isTargetPartitionOwner()) {
+            // we do not adjust the query here
+            return runQueryOnGivenPartition(query, target);
+        }
+        throw new IllegalArgumentException("Illegal target " + query);
+    }
+
+
+    // query thread first, fallback to partition thread
+    private Result runQueryOnLocalPartitions(Query query) {
         Collection<Integer> mutablePartitionIds = getLocalPartitionIds();
 
-        QueryResult result = doRunQueryOnQueryThreads(mapName, predicate, iterationType, mutablePartitionIds, LOCAL_MEMBER);
+        Result result = doRunQueryOnQueryThreads(query, mutablePartitionIds, Target.LOCAL_NODE);
         if (isResultFromAnyPartitionMissing(mutablePartitionIds)) {
-            doRunQueryOnPartitionThreads(mapName, predicate, iterationType, mutablePartitionIds, result);
+            doRunQueryOnPartitionThreads(query, mutablePartitionIds, result);
         }
 
         return result;
     }
 
     // query thread first, fallback to partition thread
-    @Override
-    public QueryResult runQueryOnAllPartitions(
-            String mapName, Predicate predicate, IterationType iterationType) {
+    private Result runQueryOnAllPartitions(Query query) {
         Collection<Integer> mutablePartitionIds = getAllPartitionIds();
```

Comment:
out of the scope for this PR, but here you could use bitset instead of collection of integers.